src/dist/component/package.rs

```diff
@@ -214,155 +158,272 @@ impl<'a> TarPackage<'a> {
+        // code was inherited from interactions with sudo that are best
+        // addressed outside of rustup (run with an appropriate effective uid).
+        // THAT SAID: If regressions turn up immediately post release this code -
+        // https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a8549057f0827bf3a068d8917256765a
+        // is a translation of the prior helper function into an in-iterator
+        // application.
+        let tar_mode = entry.header().mode().ok().unwrap();
+        // That said, the tarballs that are shipped way back have single-user
+        // permissions:
+        // -rwx------ rustbuild/rustbuild  ..... release/test-release.sh
+        // so we should normalise the mode to match the previous behaviour users
+        // may be expecting where the above file would end up with mode 0o755
+        let u_mode = tar_mode & 0o700;
+        let g_mode = (u_mode & 0o0500) >> 3;
+        let o_mode = g_mode >> 3;
+        let mode = u_mode | g_mode | o_mode;
+
+        let mut item = match kind {
+            EntryType::Directory => {
+                directories.insert(full_path.to_owned(), DirStatus::Pending(Vec::new()));
+                Item::make_dir(full_path, mode)
+            }
+            EntryType::Regular => {
+                let mut v = Vec::with_capacity(size as usize);
+                entry.read_to_end(&mut v)?;
+                Item::write_file(full_path, v, mode)
+            }
+            _ => return Err(ErrorKind::UnsupportedKind(format!("{:?}", kind)).into()),
+        };
+        budget.claim(&item);
+
+        let item = loop {
+            // Create the full path to the entry if it does not exist already
+            if let Some(parent) = item.full_path.to_owned().parent() {
+                match directories.get_mut(parent) {
+                    None => {
+                        // Tar has item before containing directory
+                        // Complain about this so we can see if these exist.
+                        eprintln!(
+                            "Unexpected: missing parent '{}' for '{}'",
+                            parent.display(),
+                            entry.path()?.display()
+                        );
+                        directories.insert(parent.to_owned(), DirStatus::Pending(vec![item]));
+                        item = Item::make_dir(parent.to_owned(), 0o755);
+                        // Check the parent's parent
+                        continue;
+                    }
+                    Some(DirStatus::Exists) => {
+                        break item;
+                    }
+                    Some(DirStatus::Pending(pending)) => {
+                        // Parent dir is being made, take next item from tar
+                        pending.push(item);
+                        continue 'entries;
+                    }
                 }
+            } else {
+                unreachable!();
```

Comment:
A comment saying why this is unreachable (or ought to be) would be good.  If it's truly unreachable then just drop the else clause, otherwise a panic would be handy.  I *think* it's truly unreachable given the shape of our tars, but a panic might be clearer, e.g. `panic!("Unexpected top level entry in tarball")` or whatever this case is.